### ISSUE TITLE
Configuration through environment

### ISSUE THREAD

**Issue by andyshinn** [containous/traefik#7]
Any plans to do configuration through environment? I'd love the ability to configure with something like:

``` shell
docker run \
  -e DOCKER_ENDPOINT="unix:///tmp/docker.sock" \
  -e DOCKER_DOMAIN="docker.localhost" \
  -e DOCKER_WATCH=true \
  -v /var/run/docker.sock:/tmp/docker.sock \
  emilevauge/traefik
```

**Comment by vdemeester** [comment]
@andyshinn cool idea yeah, it's kinda in our head to add a possibility to configure the docker image in some way :wink:

**Comment by EmileVauge** [comment]
Hi
It's in my roadmap to integrate https://github.com/spf13/viper.

**Comment by andyshinn** [comment]
Oh yea, Viper looks like a cool solution. Would definitely help!

**Comment by flrngel** [comment]
:+1:

**Comment by BrianAdams** [comment]
+1

**Comment by stongo** [comment]
+1

Does the configuration file currently inject environment variables?
Also let me know if you want help on this issue. It's something I would rather see sooner than later so willing to contribute.

**Comment by emilevauge** [comment]
Could you use flags instead?

**Comment by stongo** [comment]
Sure I can add flags for all config items instead of using ENV vars.
Would be glad to submit PR

**Comment by emilevauge** [comment]
Flags are already here in fact :)

**Comment by stongo** [comment]
For all settings found in config file? Is there a flag to disable the
config file requirement?

**Comment by stongo** [comment]
Flags still don't give the same advantage as environment variables in a container runtime for passing secrets and dynamic configuration

**Comment by emilevauge** [comment]
Right, what do you mean by dynamic configuration?

**Comment by stongo** [comment]
Anything environment specific where traefik should find itself scheduled.

Are all config items currently flags and can a config file requirement be
disabled?

**Comment by emilevauge** [comment]
@stongo: almost everything can be set using flags.
You can just give traefik an empty `traefik.toml` file and configure it using flags.
FYI, we are in the process of refactoring flags in order to simplify their maintenance.

**Comment by stongo** [comment]
Great to hear it's on the roadmap! Thanks for the advice

**Comment by RX14** [comment]
I find the deeply-nested TOML files to be really unreadable, so YAML or HCL support with viper would be brilliant.

**Comment by gjonespf** [comment]
Key use cases for viper et al for me I think would be doing things like:

- Turning on and off acme per frontend
- Specifying auth redirects per frontend
- Specifying SSL key/cert files per frontend

looking at trying to do this atm, and it looks quite difficult as is. Not sure if consul/etcd stores support key/cert changes, and ACME looks like config only from what I can see - or perhaps I've just not seen the right example yet.

So keen to see this in the roadmap.  Where is the roadmap, btw? ;)

**Comment by errm** [comment]
As I understand we have switched to https://github.com/containous/staert, I think this issue could now be closed? right @emilevauge?

**Comment by cocap10** [comment]
So far, environment variables are not handled by [https://github.com/containous/staert](https://github.com/containous/staert). But it could be a nice improvement ... WDYT @emilevauge ?

**Comment by stealthybox** [comment]
^
Environment variables are a very good way of persisting simple configuration information across a cluster.
Swarm Mode allows you to update the environment variables of existing services.
This would be a really great addition to `staert`.

**Comment by stealthybox** [comment]
would `staert` be able to support YAML inputs?

I agree with @RX14's comment on YAML or HCL being more readable than TOML in many cases.
YAML / JSON config file options would definitely make the file backend more approachable.

**Comment by RX14** [comment]
The only reason i'm not using traefik is because the examples in https://docs.traefik.io/toml/ - especially as you go further down the page - hurt my eyes to read.

TOML doesn't strike me as a format which handles nesting very well. I'm sure it looks beautiful for non-nested files but I have to say if I have to configure 10 sites using that i'm going to want to gouge my eyes out.

**Comment by InAnimaTe** [comment]
I completely and utterly agree with @RX14 

In terms of usability and inviting stature to new users, I think it would be much cleaner to adopt something like which could handle yaml or easier formats (similar to the [Caddyfile](https://caddyserver.com/docs/caddyfile)).

I think ideally, a lot of us in k8s land would want to use a configset which when updated, sees Traefik automatically reload configuration.

**Comment by errm** [comment]
Agreed, having traefik loading confine directly from a k8s config map would be nice, could you open a new issue for that @InAnimaTe ?

**Comment by InAnimaTe** [comment]
@errm done. https://github.com/containous/traefik/issues/1188

**Comment by mrmachine** [comment]
I was excited to find traefik as a potential replacement for the now deprecated dockercloud-haproxy, until I saw that I need to provide a config file and even worse, it's TOML. My eyes are bleeding.

**Comment by Yajo** [comment]
I felt the same, but then I discovered that every option you can set in the toml file, you can set it by a CLI option. It's not so beautiful, but after all it lets you basically the same benefits as those you wanted from the env variables.

**Comment by DanielJoyce** [comment]
The config file format is seriously chatty and bad, with remplating. Our plan is to use remarshaller to munge stuff together then spit out toml. :/

**Comment by Ashtonian** [comment]
An example i found digging in the docs, in case anyone has trouble finding it. https://docs.traefik.io/user-guide/cluster-docker-consul/

```yaml
version: "3"
services:
  traefik:
    image: traefik:1.5
    command:
      - "--api"
      - "--entrypoints=Name:http Address::80 Redirect.EntryPoint:https"
      - "--entrypoints=Name:https Address::443 TLS"
      - "--defaultentrypoints=http,https"
      - "--acme"
      - "--acme.storage=/etc/traefik/acme/acme.json"
      - "--acme.entryPoint=https"
      - "--acme.httpChallenge.entryPoint=http"
      - "--acme.onHostRule=true"
      - "--acme.onDemand=false"
      - "--acme.email=contact@mydomain.ca"
      - "--docker"
      - "--docker.swarmMode"
      - "--docker.domain=mydomain.ca"
      - "--docker.watch"

```

**Comment by dantebarba** [comment]
> An example found digging in the docs, using flags in case anyone has trouble finding it. https://docs.traefik.io/v1.7/user-guide/cluster-docker-consul/
> 
> ```yaml
> version: "3"
> services:
>   traefik:
>     image: traefik:1.7
>     command:
>       - "--api"
>       - "--entrypoints=Name:http Address::80 Redirect.EntryPoint:https"
>       - "--entrypoints=Name:https Address::443 TLS"
>       - "--defaultentrypoints=http,https"
>       - "--acme"
>       - "--acme.storage=/etc/traefik/acme/acme.json"
>       - "--acme.entryPoint=https"
>       - "--acme.httpChallenge.entryPoint=http"
>       - "--acme.onHostRule=true"
>       - "--acme.onDemand=false"
>       - "--acme.email=contact@mydomain.ca"
>       - "--docker"
>       - "--docker.swarmMode"
>       - "--docker.domain=mydomain.ca"
>       - "--docker.watch"
> ```

What happens if I just want to replace acme.email for example and keep using the rest of the toml file?. Can I just set --acme.email and it will override the toml?

**Comment by traefiker** [comment]
Closed by #4935.
